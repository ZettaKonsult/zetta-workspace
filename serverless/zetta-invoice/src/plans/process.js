import { incrementToNextLowerBound } from 'date-primitive-utils';

import { getDbTable } from '../util/database';

import { listAllPlansToProcess } from './list';

const table = getDbTable({ name: 'Plans' });

export default async ({ db }) => {
  const plans = await listAllPlansToProcess({ db });

  /*
    DO SOME KIND OF PROCESS TO SEND INVOICE TO ALL PLANS
  */

  const result = await updateNextProcess({ db, plans });
  console.log(result);
};

const updateNextProcess = async ({ db, plans }) => {
  const updatePromise = plans.map(item => {
    return db('update', {
      TableName: table,
      Key: {
        companyCustomerId: item.companyCustomerId,
        id: item.id,
      },
      ReturnValues: 'ALL_NEW',
      UpdateExpression: 'set #epochNextProcess = :epochNextProcess',
      ExpressionAttributeNames: {
        '#epochNextProcess': 'epochNextProcess',
      },
      ExpressionAttributeValues: {
        ':epochNextProcess': incrementToNextLowerBound(
          item.epochNextProcess,
          item.intervalCount
        ),
      },
    });
  });
  const result = Promise.all(updatePromise);
  return result;
};
