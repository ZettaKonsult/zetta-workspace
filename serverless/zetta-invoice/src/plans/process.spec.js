import { mapPlansToRecipients, convertToInvoice } from './process';

describe('converts array to map', () => {
  const plans = [
    { id: 'p1', name: 'pTest1', recipientIds: ['r1'], companyCustomerId: 'c1' },
    { id: 'p2', name: 'pTest2', recipientIds: ['r1'], companyCustomerId: 'c1' },
    { id: 'p3', name: 'pTest3', recipientIds: ['r2'], companyCustomerId: 'c1' },
    { id: 'p4', name: 'pTest4', recipientIds: ['r2'], companyCustomerId: 'c2' },
  ];

  it('converts a mapping between customer, recipient and plans', () => {
    const result = mapPlansToRecipients({ plans });

    expect(result).toEqual({
      r1: {
        c1: ['p1', 'p2'],
      },
      r2: {
        c1: ['p3'],
        c2: ['p4'],
      },
    });
  });

  it('converts map to invoice', () => {
    const result = mapPlansToRecipients({ plans });
    const invoice = convertToInvoice({ plans, plansToRecipientsMap: result });

    expect(invoice).toEqual([
      {
        companyCustomerId: 'c1',
        invoiceRows: [
          {
            description: 'pTest1',
            price: 0,
            tax: 0.25,
            unit: 1,
          },
          {
            description: 'pTest2',
            price: 0,
            tax: 0.25,
            unit: 1,
          },
        ],
        recipientId: 'r1',
      },
      {
        companyCustomerId: 'c1',
        invoiceRows: [
          {
            description: 'pTest3',
            price: 0,
            tax: 0.25,
            unit: 1,
          },
        ],
        recipientId: 'r2',
      },
      {
        companyCustomerId: 'c2',
        invoiceRows: [
          {
            description: 'pTest4',
            price: 0,
            tax: 0.25,
            unit: 1,
          },
        ],
        recipientId: 'r2',
      },
    ]);
  });
});
