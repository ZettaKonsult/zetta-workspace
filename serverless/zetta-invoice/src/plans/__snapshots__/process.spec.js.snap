// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts array to map converts a mapping between customer, recipient and plans 1`] = `
Object {
  "r1": Object {
    "c1": Array [
      "p1",
      "p2",
    ],
  },
  "r2": Object {
    "c1": Array [
      "p3",
    ],
    "c2": Array [
      "p4",
    ],
  },
}
`;

exports[`converts array to map converts map to invoice 1`] = `
Array [
  Object {
    "companyCustomerId": "c1",
    "invoiceRows": Array [
      Object {
        "description": "pTest1",
        "price": 0,
        "tax": 0.25,
        "unit": 1,
      },
      Object {
        "description": "pTest2",
        "price": 0,
        "tax": 0.25,
        "unit": 1,
      },
    ],
    "recipientId": "r1",
  },
  Object {
    "companyCustomerId": "c1",
    "invoiceRows": Array [
      Object {
        "description": "pTest3",
        "price": 0,
        "tax": 0.25,
        "unit": 1,
      },
    ],
    "recipientId": "r2",
  },
  Object {
    "companyCustomerId": "c2",
    "invoiceRows": Array [
      Object {
        "description": "pTest4",
        "price": 0,
        "tax": 0.25,
        "unit": 1,
      },
    ],
    "recipientId": "r2",
  },
]
`;
